toxcore/tox.c

```diff
@@ -309,20 +312,46 @@ static void tox_conference_peer_list_changed_handler(Messenger *m, uint32_t conf
 static void tox_friend_lossy_packet_handler(Messenger *m, uint32_t friend_number, uint8_t packet_id,
         const uint8_t *data, size_t length, void *user_data)
 {
+    if (!data) {
+        // no error handling?
+        return;
+    }
+
+    if (length < 1) {
+        // no error handling?
+        return;
+    }
+
     struct Tox_Userdata *tox_data = (struct Tox_Userdata *)user_data;
 
-    if (tox_data->tox->friend_lossy_packet_callback != nullptr) {
-        tox_data->tox->friend_lossy_packet_callback(tox_data->tox, friend_number, data, length, tox_data->user_data);
+    if (tox_data->tox->friend_lossy_packet_callback_per_pktid[data[0]] != nullptr) {
+        unlock(tox_data->tox);
+        tox_data->tox->friend_lossy_packet_callback_per_pktid[data[0]](tox_data->tox, friend_number, data, length,
+                tox_data->user_data);
+        lock(tox_data->tox);
     }
 }
 
 static void tox_friend_lossless_packet_handler(Messenger *m, uint32_t friend_number, uint8_t packet_id,
         const uint8_t *data, size_t length, void *user_data)
 {
+    if (!data) {
+        // no error handling?
+        return;
+    }
+
+    if (length < 1) {
+        // no error handling?
+        return;
+    }
+
     struct Tox_Userdata *tox_data = (struct Tox_Userdata *)user_data;
 
-    if (tox_data->tox->friend_lossless_packet_callback != nullptr) {
-        tox_data->tox->friend_lossless_packet_callback(tox_data->tox, friend_number, data, length, tox_data->user_data);
+    if (tox_data->tox->friend_lossless_packet_callback_per_pktid[data[0]] != nullptr) {
+        unlock(tox_data->tox);
+        tox_data->tox->friend_lossless_packet_callback_per_pktid[data[0]](tox_data->tox, friend_number, data, length,
+                tox_data->user_data);
+        lock(tox_data->tox);
```

Comment:
Make a separate PR for all the unlock/lock and non-recursive mutex changes.

Reply:
that slipped in there by accident